### PR TITLE
Add dxGantt widget

### PR DESCRIPTION
**Pull request by kruglikovstepan**

### CODE REVIEW

**Review by maxerokh — DISMISSED** [review]

File: build/gulp/vendor.js [hunk 1]
```diff
@@ -41,12 +41,15 @@ const JS_VENDORS = [
...
+    },
+    {
+        path: '/devexpress-gantt/dist/dx-@(gantt|gantt.min).js'
     }
 ];
 
 const CSS_VENDORS = [
     {
-        path: '/devexpress-diagram/dist/dx-@(diagram|diagram.min).css'
+        path: '/devexpress-gantt/dist/dx-@(gantt|gantt.min).css'
```
Comment (vconst) [inline build/gulp/vendor.js:55]: Why did diagram replace to gantt?

File: js/ui/gantt/ui.gantt.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import { extend } from '../../core/utils/extend';
+
+const GANTT_CLASS = 'dx-gantt';
+const GANTT_SPLITTER_CLASS = 'dx-gantt-splitter';
+const GANTT_VIEW_CLASS = 'dx-gantt-view';
+
+class Gantt extends Widget {
+    _init() {
+        super._init();
+    }
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.js]: There is no logic in this _init overload. Remote it.

File: js/ui/gantt/ui.gantt.view.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    getGanttTickSize() {
+        return { width: 100, height: 34 };
+    }
+    getGanttContainer() {
+        return this.$element().get(0);
+    }
+    getGanttViewStartDate() {
+        let min = new Date();
+        const tasks = this.getGanttTasksData();
+        tasks.forEach(t => {
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.view.js]: Using Array,reduce is better to find min value

File: ts/dx.all.d.ts [hunk 4]
```diff
@@ -2673,6 +2680,22 @@ declare module DevExpress.ui {
...
+    /** @name dxGantt.Options */
+    export interface dxGanttOptions extends WidgetOptions<dxGantt> {
+        /** @name dxGantt.Options.dependenciesDataSource */
+        dependenciesDataSource?: any;
+        /** @name dxGantt.Options.resourceAssignmentsDataSource */
+        resourceAssignmentsDataSource?: any;
+        /** @name dxGantt.Options.resourcesDataSource */
+        resourcesDataSource?: any;
+        /** @name dxGantt.Options.taskDataSource */
+        taskDataSource?: any;
```
Comment (vconst) [inline ts/dx.all.d.ts]: Why is task prefix singular but dependencies, resourceAssignments and resources are plural

Comment (vconst) [reply ts/dx.all.d.ts]: Also, "any" type is not good for data sources.

File: js/ui/gantt/ui.gantt.view.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            if(t.start < min) {
+                min = t.start;
+            }
+        });
+        return min;
+    }
+    getGanttViewEndDate() {
+        let max = new Date();
+        const tasks = this.getGanttTasksData();
+        tasks.forEach(t => {
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.view.js]: Using Array,reduce is better to find max value

File: js/ui/gantt/ui.gantt.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        });
+    }
+
+    _optionChanged(args) {
+        switch(args.name) {
+            case 'taskDataSource':
+            case 'dependenciesDataSource':
+            case 'resourcesDataSource':
+            case 'resourceAssignmentsDataSource':
+                break;
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.js:142]: Why is no logic here?

File: js/ui/gantt/ui.gantt.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
+import $ from '../../core/renderer';
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.js]: Use everywhere double quotes because devextreme widgets use them.

File: js/ui/gantt/ui.gantt.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            scrollable.off('scroll', (e) => { this._onTreeListScroll(e); });
+            scrollable.on('scroll', (e) => { this._onTreeListScroll(e); });
+        }
+    }
+
+    _getDefaultOptions() {
+        return extend(super._getDefaultOptions(), {
+            /**
+            * @name dxGanttOptions.taskDataSource
+            * @type object
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.js]: The object type is bad for datasource. Mostly our widgets have the following type for datasource:
* @type Array<Object>|DataSource|DataSourceOptions

**Review by vconst — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets/gantt.tests.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        this.instance = this.$element.dxGantt("instance");
+    }
+};
+
+QUnit.module("Gantt", moduleConfig, () => {
+    test("should render treeList", (assert) => {
+        const treeListElement = this.$element.find(TREELIST_SELECTOR);
+        assert.ok(treeListElement.length === 1);
+    });
+});
```
Comment (vconst) [inline testing/tests/DevExpress.ui.widgets/gantt.tests.js:101]: Add tests for each Gantt option initialization and for each option changing

**Review by vconst — CHANGES_REQUESTED** [review]

File: js/ui/gantt/ui.gantt.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+            */
+            taskDataSource: null,
+            /**
+            * @name dxGanttOptions.dependenciesDataSource
+            * @type object
+            * @default null
+            */
+            dependenciesDataSource: null,
+            /**
+            * @name dxGanttOptions.resourcesDataSource
```
Comment (RomanTsukanov) [inline js/ui/gantt/ui.gantt.js]: Do we need that "dataSource" postfix in every option name? IMHO, simple names like `tasks`, `dependencies`, `resources`, and `resourceAssignments` (or `resourceAssignees`, see my comment below) are much better

File: js/ui/gantt/ui.gantt.js [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+            * @type object
+            * @default null
+            */
+            resourcesDataSource: null,
+            /**
+            * @name dxGanttOptions.resourceAssignmentsDataSource
```
Comment (RomanTsukanov) [inline js/ui/gantt/ui.gantt.js]: Does this data source contain assignees, like employees?

**Review by RomanTsukanov — DISMISSED** [review]
Doccomments LGTM

File: js/ui/gantt/ui.gantt.view.js [hunk 12]
```diff
@@ -44,36 +44,24 @@ export class GanttView extends Widget {
...
     getGanttViewStartDate() {
-        let min = new Date();
         const tasks = this.getGanttTasksData();
-        tasks.forEach(t => {
-            if(t.start < min) {
-                min = t.start;
-            }
-        });
-        return min;
+        return tasks.reduce((min, t) => t.start < min ? t.start : min, tasks[0].start);
```
Comment (vconst) [inline js/ui/gantt/ui.gantt.view.js]: Null ref if tasks.length is 0. May be, use new Date() instead of tasks[0].start

**Review by vconst — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets/gantt.tests.js (hunk 9 above)
Comment (kruglikovstepan) [reply testing/tests/DevExpress.ui.widgets/gantt.tests.js:101]: Tests will be added after the optionsChanged method implementation

Comment (vconst) [reply testing/tests/DevExpress.ui.widgets/gantt.tests.js:101]: Are you not using TDD? Tests about options initialization can be added now.

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by vconst — APPROVED** [review]

**Pull request merged**